gpflow/utilities/utilities.py

```diff
@@ -265,7 +265,16 @@ def _set_by_name_index(parent: object, value: Any, attr_str: str, index_str: Uni
         index = int(index_str)
         attr = getattr(parent, attr_str)
         attr[index] = value
+        # tensorflow's __setattr__ override does not contain a check for the case where
+        # an attribute holding a trackable object (e.g. a Variable) is being reassigned
+        # a non-trackable object (e.g. a constant). Therefore, tensorflow still stores
+        # internal references to the trackable object after reassignment, which can lead
+        # to various problems, for example https://github.com/GPflow/GPflow/pull/1338
+        # By calling delattr first we are forcing tensorflow to remote its internal reference.
+        delattr(parent, attr_str)
```

Comment:
should this not delete getattr(parent, attr_str)[index] instead ?... :thinking:

Reply:
I am a bit concerned about the performance hit of deleting list elements. If we are deleting in the beginning of the list, all the elements after it will be moved, same when inserting into the list. This way, by deleting the entire list, this is (in theory, I do not know about tensorflow internals to be honest) an O(1) operation, as Python attributes are stored in a dict.